### PR TITLE
472: fix TMB option warning

### LINKED ISSUE #472: mmrm 0.3.13's first fit is different than subsequent ones
**Summary**
mmrm 0.3.13 gives a different result on the first call than subsequent ones. Note that with TMB 1.9.15 and mmrm 0.3.12 this does not happen (the result is the same every time). 

I've been able to narrow it down to the TMB::config call in the warning about optimize.instantly (https://github.com/openpharma/mmrm/blob/main/R/utils.R#L536 - also, if I manually set optimize.instantly to 1, the warning does not trigger?), so this seems to be a side effect of calling `TMB::config` to get the configuration even without changing the config? Reproducible example that allows for switching mmrm versions easily below:

```r
mmrm_version <- "0.3.13" # "0.3.12" for comparison
mmrm_libpath <- paste0(".libs-", mmrm_version)

if(!dir.exists(mmrm_libpath)){
  dir.create(mmrm_libpath)
}
.libPaths(c(normalizePath(mmrm_libpath), .libPaths()))

if (!dir.exists(paste0(mmrm_libpath, "/mmrm"))) {
  remotes::install_version("mmrm", version = mmrm_version, type="source", repos="https://stat.ethz.ch/CRAN", dependencies="never")
}

library(mmrm)
library(digest)
options(digits = 22)

res <- character()
cat(as.character(packageVersion("mmrm")), "\n")
for (i in 1:5) {
  set.seed(123)
  mmrmFit <- mmrm(
    formula = FEV1 ~ RACE + SEX + ARMCD * AVISIT + us(AVISIT | USUBJID),
    data = fev_data
  )
  res[i] <- digest(coef(mmrmFit))
  cat("Res ", i, ": ", res[i], "\n")
}
```

Output for mmrm 0.3.13:
```r
> source("~/.active-rstudio-document")
0.3.13 
Res  1 :  a005171052e35c971e920321dea782f6 
Res  2 :  903104ec4747c253d773e48e6ab96b32 
Res  3 :  903104ec4747c253d773e48e6ab96b32 
Res  4 :  903104ec4747c253d773e48e6ab96b32 
Res  5 :  903104ec4747c253d773e48e6ab96b32
```

Output for mmrm 0.3.12:
```r
> source("~/.active-rstudio-document")
0.3.12 
Res  1 :  903104ec4747c253d773e48e6ab96b32 
Res  2 :  903104ec4747c253d773e48e6ab96b32 
Res  3 :  903104ec4747c253d773e48e6ab96b32 
Res  4 :  903104ec4747c253d773e48e6ab96b32 
Res  5 :  903104ec4747c253d773e48e6ab96b32 
```

**OS / Environment**
Multiple (R 4.4 on ARM64 / MacOS, R 4.3.1 on RHEL 7)

@danielinteractive @kaskr any ideas?

### PR DESCRIPTION
**Pull request by danielinteractive**
closes #472

### CODE REVIEW

**Comment by danielinteractive** [comment]
@kaskr @luwidmer if you can have quick look that would be great, thank you 😃

**Comment by kaskr** [comment]
@danielinteractive This looks right to me.

File: NEWS.md [hunk 1]
```diff
@@ -2,6 +2,12 @@
 
 ### Bug Fixes
 
+- Previously, when the tape optimizer from `TMB` was switched on a warning would be given, instructing to turn off the tape optimizer. However, this is not necessary for reproducible results. Instead now it is checked whether the deterministic hash for the TMB tape optimizer is used, and a warning is issued otherwise.
```
Comment (luwidmer) [inline NEWS.md:5]: Might make sense to say that this resulted in the first fit being fit with a different optimizer and thus being different?

**Review by luwidmer — COMMENTED** [review]
Can confirm this fixes the first-fit bug:

```r
> source("~/.active-rstudio-document")
0.3.13.9000 
Res  1 :  903104ec4747c253d773e48e6ab96b32 
Res  2 :  903104ec4747c253d773e48e6ab96b32 
Res  3 :  903104ec4747c253d773e48e6ab96b32 
Res  4 :  903104ec4747c253d773e48e6ab96b32 
Res  5 :  903104ec4747c253d773e48e6ab96b32 
```


Also, I see the warnings trigger on TMB 1.9.14 as well as 1.9.15 with non-deterministic hash.

**Review by luwidmer — APPROVED** [review]

**Pull request merged**